//
// PROJECT:         Aspia
// FILE:            system_info/category_ata.proto
// LICENSE:         Mozilla Public License Version 2.0
// PROGRAMMERS:     Dmitry Chapyshev (dmitry@aspia.ru)
//

syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package aspia.proto;

message ATA
{
    enum BusType
    {
        BUS_TYPE_UNKNOWN             = 0;
        BUS_TYPE_SCSI                = 1;
        BUS_TYPE_ATAPI               = 2;
        BUS_TYPE_ATA                 = 3;
        BUS_TYPE_IEEE1394            = 4;
        BUS_TYPE_SSA                 = 5;
        BUS_TYPE_FIBRE               = 6;
        BUS_TYPE_USB                 = 7;
        BUS_TYPE_RAID                = 8;
        BUS_TYPE_ISCSI               = 9;
        BUS_TYPE_SAS                 = 10;
        BUS_TYPE_SATA                = 11;
        BUS_TYPE_SD                  = 12;
        BUS_TYPE_MMC                 = 13;
        BUS_TYPE_VIRTUAL             = 14;
        BUS_TYPE_FILE_BACKED_VIRTUAL = 15;
    }

    enum TransferMode
    {
        TRANSFER_MODE_UNKNOWN       = 0;
        TRANSFER_MODE_PIO           = 1;
        TRANSFER_MODE_PIO_DMA       = 2;
        TRANSFER_MODE_ULTRA_DMA_133 = 3;
        TRANSFER_MODE_ULTRA_DMA_100 = 4;
        TRANSFER_MODE_ULTRA_DMA_66  = 5;
        TRANSFER_MODE_ULTRA_DMA_44  = 6;
        TRANSFER_MODE_ULTRA_DMA_33  = 7;
        TRANSFER_MODE_ULTRA_DMA_25  = 8;
        TRANSFER_MODE_ULTRA_DMA_16  = 9;
        TRANSFER_MODE_SATA_600      = 10;
        TRANSFER_MODE_SATA_300      = 11;
        TRANSFER_MODE_SATA_150      = 12;
    }

    enum Features
    {
        FEATURE_UNKNOWN                       = 0;
        FEATURE_48BIT_LBA                     = 1;
        FEATURE_ADVANCED_POWER_MANAGEMENT     = 2;
        FEATURE_AUTOMATIC_ACOUSTIC_MANAGEMENT = 4;
        FEATURE_SMART                         = 8;
        FEATURE_SMART_ERROR_LOGGING           = 16;
        FEATURE_SMART_SELF_TEST               = 32;
        FEATURE_STREAMING                     = 64;
        FEATURE_GENERAL_PURPOSE_LOGGING       = 128;
        FEATURE_SECURITY_MODE                 = 256;
        FEATURE_POWER_MANAGEMENT              = 512;
        FEATURE_WRITE_CACHE                   = 1024;
        FEATURE_READ_LOCK_AHEAD               = 2048;
        FEATURE_HOST_PROTECTED_AREA           = 4096;
        FEATURE_RELEASE_INTERRUPT             = 8192;
        FEATURE_POWER_UP_IN_STANDBY           = 16384;
        FEATURE_DEVICE_CONFIGURATION_OVERLAY  = 32768;
        FEATURE_SERVICE_INTERRUPT             = 65536;
        FEATURE_NATIVE_COMMAND_QUEUING        = 131072;
        FEATURE_TRIM                          = 262144;
    }

    message Item
    {
        string model_number        = 1;
        string serial_number       = 2;
        string firmware_revision   = 3;
        BusType bus_type           = 4;
        TransferMode transfer_mode = 5;
        uint32 rotation_rate       = 6;
        uint64 drive_size          = 7;
        uint32 buffer_size         = 8;
        uint32 multisectors        = 9;
        uint32 ecc_size            = 10;
        bool is_removable          = 11;
        uint64 cylinders_number    = 12;
        uint32 tracks_per_cylinder = 13;
        uint32 sectors_per_track   = 14;
        uint32 bytes_per_sector    = 15;
        uint32 heads_number        = 16;
        uint64 supported_features  = 17;
        uint64 enabled_features    = 18;
    }

    repeated Item item = 1;
}
